### DISCUSSION TITLE
Index built on GPU, Searching on this index is not on GPU

### DISCUSSION THREAD

**Discussion by xiaojuan8** [milvus-io/milvus#15735] [Q&A and General discussion]
Hi, I have built an index with GPU-on, and index was save on milvus. I call search() on this index, GPU is not working, and CPU is very busy(about 700%). How can milvus do search on GPU?

**Comment by yhmo** [comment]
To search on GPU:
Firstly, the index can be computed by GPU, currently only the IVF series and brute-force search can be executed by GPU in v1.x.
Secondly, there is a threshold "gpu_search_threshold" defined in the server_config.yaml
```
gpu:
  enable: true
  gpu_search_threshold: 1000
```
If the nq of search() is less than "gpu_search_threshold", milvus will use CPU; otherwise use GPU.
The default value is 1000, you can change it to 1, then milvus will always use GPU to search.

**Reply by xiaojuan8** [reply]
Hi, thanks, I use IVFPQ index, and milvus v1.1, which does support GPU building index.
I will try to set gpu_search_threshold to 1, thanks so much.

**Comment by yhmo** [comment]
For what is "nq":
"nq" is the vectors count you input to the parameter "query_records".
```
    q_vectors = [[random.random() for _ in range(COLLECTION_DIMENSION)] for _ in range(NQ)]
    search_future = MILVUS.search(collection_name=name, query_records=q_vectors, top_k=TOPK, params=SEARCH_PARAM)
```

**Reply by xiaojuan8** [reply]
Hi, I set nq threshold for GPU searching. And GPU is working while searching. The problem is it is so slow, but GPU is not utilized to 100% or much as they can. only 20% of GPU power is used.
![image](https://user-images.githubusercontent.com/74544828/155491928-45ba474b-f928-4eea-80ab-3348693ecc05.png)

and one search elapse time is much higher than CPU.
they search on the same index, (same nq,topk,nprobe etc, client is all the same on GPU and CPU searching)
GPU
![image](https://user-images.githubusercontent.com/74544828/155492051-142f888c-ea72-4ddd-aff2-4acf855e3877.png)

CPU
![image](https://user-images.githubusercontent.com/74544828/155492180-50700ed6-1ec3-40c7-af3f-a434add6b93d.png)

**Reply by xiaojuan8** [reply]
server_config.yml
GPU config:
![image](https://user-images.githubusercontent.com/74544828/155492838-a8141339-747b-49a8-9ed7-334ac06b655f.png)

**Comment by yhmo** [comment]
Try set the gpu.cache.enable=true
```
gpu:
  enable: true
  cache.enable: true
```

The gpu.cache.enable is false by default, each search will copy segment data from CPU memory to GPU memory one by one.
If the gpu.cache.enable = true. milvus will try keep segments in GPU memory.

By the way. Totally how many vectors in the collection? And the dimension?
